src/game/client/draw/w3dmodeldraw.cpp

```diff
@@ -1154,3 +1178,1890 @@ void W3DModelDrawModuleData::Parse_Condition_State(INI *ini, void *instance, voi
+        info.m_allowUpdates = false;
+        info.m_allowWorldAlign = true;
+        info.m_type = (ShadowType)info.m_thing->Get_Shadow_Type();
+        info.m_sizeX = info.m_thing->Get_Shadow_Size_X();
+        info.m_sizeY = info.m_thing->Get_Shadow_Size_Y();
+        info.m_offsetX = info.m_thing->Get_Shadow_Offset_X();
+        info.m_offsetY = info.m_thing->Get_Shadow_Offset_Y();
+        m_shadow = g_theW3DShadowManager->Add_Shadow(m_renderObject, &info, nullptr);
+
+        if (m_shadow != nullptr) {
+            m_shadow->Enable_Shadow_Invisible(m_fullyObscuredByShroud);
+
+            if (m_renderObject->Is_Hidden() || !m_hidden) {
+                m_shadow->Enable_Shadow_Render(false);
+            }
+        }
+    }
+}
+
+void W3DModelDraw::Set_Shadows_Enabled(bool enable)
+{
+    if (m_shadow != nullptr) {
+        m_shadow->Enable_Shadow_Render(enable);
+    }
+
+    m_hidden = enable;
+}
+
+#ifdef GAME_DEBUG_STRUCTS
+void W3DModelDraw::Gather_Draw_Stats(DebugDrawStats *stats)
+{
+    Gather_Draw_Stats_For_Render_Object(stats, m_renderObject);
+
+    if (m_shadow != nullptr) {
+        m_shadow->Gather_Draw_Stats(stats);
+    }
+}
+#endif
+
+void W3DModelDraw::Set_Fully_Obscured_By_Shroud(bool obscured)
+{
+    if (m_fullyObscuredByShroud != obscured) {
+        m_fullyObscuredByShroud = obscured;
+
+        if (m_shadow != nullptr) {
+            m_shadow->Enable_Shadow_Invisible(m_fullyObscuredByShroud);
+        }
+
+        if (m_decalShadow != nullptr) {
+            m_decalShadow->Enable_Shadow_Invisible(m_fullyObscuredByShroud);
+        }
+
+        Do_Start_Or_Stop_Particle_Sys();
+    }
+}
+
+void W3DModelDraw::Adjust_Transform_Mtx(Matrix3D &transform) const
+{
+    Vector3 *offset = Get_W3D_Model_Draw_Module_Data()->Get_Attach_To_Drawable_Bone_Offset(Get_Drawable());
+
+    if (offset != nullptr) {
+        Vector3 v = transform.Rotate_Vector(*offset);
+        transform.Adjust_X_Translation(v.X);
+        transform.Adjust_Y_Translation(v.Y);
+        transform.Adjust_Z_Translation(v.Z);
+    }
+
+    if ((m_curState->m_flags & MAINTAIN_FRAME_ACROSS_STATES3) != 0) {
+        const Object *object = Get_Drawable()->Get_Object();
+
+        if (object != nullptr) {
+            float percent = object->Get_Construction_Percent();
+
+            if (percent >= 0.0f) {
+                float height = object->Get_Geometry_Info().Get_Max_Height_Above_Position();
+                transform.Translate_Z(height * percent / 100.0f - height);
+            }
+        }
+    }
+}
+
+bool Is_Animation_Complete(RenderObjClass *r)
+{
+    if (r != nullptr && r->Class_ID() == RenderObjClass::CLASSID_HLOD) {
+        return static_cast<HLodClass *>(r)->Is_Animation_Complete();
+    } else {
+        return true;
+    }
+}
+
+void W3DModelDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    Set_Pause_Animation(Get_Drawable()->Get_Should_Animate(Get_W3D_Model_Draw_Module_Data()->m_animationsRequirePower));
+    Matrix3D m;

```

Comment:
This declaration can be moved into the scope below. There is another `Matrix3D m` in this function that currently shadows this one.

Reply:
fixed.